keel-bakery-plugin/src/main/kotlin/com/netflix/spinnaker/keel/bakery/artifact/ImageHandler.kt

```diff
@@ -36,30 +39,33 @@ class ImageHandler(
           ArtifactCheckSkipped(artifact.type, artifact.name, "ActuationInProgress")
         )
       } else {
-        val latestVersion = artifact.findLatestArtifactVersion()
-        val latestBaseImageVersion = artifact.findLatestBaseAmiVersion()
-        val image = artifact.findLatestAmi()
-
-        if (image == null) {
-          log.debug("No image found for {}", artifact.name)
-        } else {
-          log.debug("Latest known {} version: {}", artifact.name, latestVersion)
-          log.debug("Latest baked {} version: {}", artifact.name, image.appVersion)
-          log.debug("Latest {} {} base AMI version: {}", artifact.vmOptions.baseOs, artifact.vmOptions.baseLabel, latestBaseImageVersion)
-          log.debug("Latest baked {} base AMI version: {}", artifact.name, image.baseAmiVersion)
-        }
-
-        val versionsDiffer = image?.appVersion != latestVersion || image?.baseAmiVersion != latestBaseImageVersion
-        val regionsDiffer = !(image?.regions ?: emptySet()).containsAll(artifact.vmOptions.regions)
-        if (image == null || versionsDiffer || regionsDiffer) {
-          log.info("Versions or regions for {} differ, launching bake", artifact.name)
-          launchBake(artifact, latestVersion)
+        val latestArtifactVersion = artifact.findLatestArtifactVersion()
+        val latestBaseAmiVersion = artifact.findLatestBaseAmiVersion()
+
+        val desired = Image(latestBaseAmiVersion, latestArtifactVersion, artifact.vmOptions.regions)
+        val current = artifact.findLatestAmi()
+        val diff = DefaultResourceDiff(desired, current)
+
+        if (diff.hasChanges()) {
+          if (current == null) {
+            log.info("No AMI found for {}", artifact.name)
+          } else {
+            log.info("Image for {} delta: {}", artifact.name, diff.toDebug())
+          }
+
+          if (diffFingerprintRepository.seen(artifact.name, diff)) {
+            log.warn("Artifact version {} and base AMI version {} were baked previously", latestArtifactVersion, latestBaseAmiVersion)
+            publisher.publishEvent(RecurrentBakeDetected(latestArtifactVersion, latestBaseAmiVersion))
+          } else {
+            launchBake(artifact, latestArtifactVersion)
+            diffFingerprintRepository.store(artifact.name, diff)
+          }
         } else {
           log.debug("Existing image for {} is up-to-date", artifact.name)
         }
 
-        if (image != null && regionsDiffer) {
-          publisher.publishEvent(ImageRegionMismatchDetected(image, artifact.vmOptions.regions))
+        if (current != null && diff.affectedRootPropertyNames == setOf("regions")) {
+          publisher.publishEvent(ImageRegionMismatchDetected(current, artifact.vmOptions.regions))
```

Comment:
Great instrumentation. We'd been missing data on the frequency of this forever. 👏